src/Build/BackEnd/Components/RequestBuilder/TargetBuilder.cs

```diff
@@ -766,5 +767,45 @@ private async Task<bool> CompleteOutstandingActiveRequests(string targetName)
 
             return false;
         }
+
+        private void ComputeAfterTargetFailures(string[] targetNames)
+        {
+            // Cache intermediate results so we don't visit a target more than once.
+            Dictionary<string, bool> resultsWithAfterTargets = new Dictionary<string, bool>();
+
+            foreach (string targetName in targetNames)
+            {
+                if (_buildResult.ResultsByTarget.ContainsKey(targetName))
+                {
+                    // Queue of targets waiting to be processed, seeded with the specific target for which we're computing AfterTargetsHaveFailed.
+                    Queue<string> targetsToCheckForAfterTargets = new Queue<string>();
+                    targetsToCheckForAfterTargets.Enqueue(targetName);
+
+                    while (targetsToCheckForAfterTargets.Count > 0)
+                    {
+                        string targetToCheck = targetsToCheckForAfterTargets.Dequeue();
+                        IList<TargetSpecification> targetsWhichRunAfter = _requestEntry.RequestConfiguration.Project.GetTargetsWhichRunAfter(targetToCheck);
+
+                        foreach (TargetSpecification afterTarget in targetsWhichRunAfter)
+                        {
+                            if (!resultsWithAfterTargets.TryGetValue(afterTarget.TargetName, out bool afterTargetFailed))
+                            {
+                                _buildResult.ResultsByTarget.TryGetValue(afterTarget.TargetName, out TargetResult result);
+                                afterTargetFailed = result != null && result.ResultCode == TargetResultCode.Failure && !result.TargetFailureDoesntCauseBuildFailure;
+
+                                resultsWithAfterTargets[afterTarget.TargetName] = afterTargetFailed;
+                                if (afterTargetFailed)
+                                {
+                                    _buildResult.ResultsByTarget[targetName].AfterTargetsHaveFailed = true;
```

Comment:
As an optimization, if you found one failing after target, why not exit the while loop prematurely?

Reply:
Great point, done!